### ISSUE TITLE
incorrect sum aggregation

### ISSUE THREAD

**Issue by hamelg** [graphite-ng/carbon-relay-ng#184]
Sometimes, aggregation sum is incorrect. Some values are missing randomly when summing.  Here is a example :

![graph](https://cloud.githubusercontent.com/assets/9417309/25662011/2dd5f710-3013-11e7-86ea-d292a827fce7.png)

I have 3 values aggregated like that :

`sum                       (.+)\.QWGTI\d+\.kubernetes\.pods\.([^.]+)\.([^.]+?)(-\d{1,2})?-[^.]{5}\.([^.]+)\.usage_rate                $1.kubernetes.namespaces.$2.pods.$3.$5.usage_rate    60   180`

svc-bootstrap-back-spbrt-4-h4n6c.usage_rate + svc-bootstrap-back-spbrt-4-0q5ee.usage_rate + svc-bootstrap-back-spbrt-4-y8jxr.usage_rate = usage_rate

you see the issue on the graph at 10:07, 10:11, 10:34 and 10:44.

Any ideas ?
How to troubleshoot ?

If I route theses metrics and doing this aggregation on a idle relay-ng instance, the issue vanishes.

**Comment by Dieterbe** [comment]
> If I route theses metrics and doing this aggregation on a idle relay-ng instance, the issue vanishes.

so in the case where the issue is visible, is your relay overloaded maybe? how is cpu usage like?

**Comment by hamelg** [comment]
The server is not loaded at all (80% idle).
To workaround the issue, we have to split the relay_ng instance in 2 instances : 
1 to collect on network, blacklist, rewrite and route
1 to only aggregate
The 2 instances are connected via localhost.
It looks like a bug because the raw data is ok, but the aggregated sum values have hicups.
Here is a example on aggregated values where we have applied the workaround at 16:30.
![screenshot_20170503_200253](https://cloud.githubusercontent.com/assets/9417309/25674623/d3666afc-303b-11e7-9d7d-a26936f0e2ec.png)

**Comment by Dieterbe** [comment]
> The server is not loaded at all (80% idle).

is that measurement across multiple cores? when the issue appears, is the relay using most of any particular core? the relay currently doesn't scale well across multiple cores, and if the 1 cpu it runs on saturates, that could explain measurements not making it into the aggregator on time. though you have a wait of 180s (3 minutes) which should be plenty. are you sure that the data comes in consistently at every minute, and doesn't come in late? (e.g. if metrics for 1 series come in a few minutes after those for the other two series, then it wouldn't be included in the aggregation, though the input data would look fine in grafana - after the fact)

**Comment by hamelg** [comment]
The server has 2 cores and it is lightly loaded : the max load never raise 0.8, the mean is near 0.4. We see the same load before and after 16:30. Initially, the wait was at 120s, but I increased it at 180s without success because I suspected the issue was related to data coming in late. If it can help, here is the configuration file where issue happen before 16:30 [carbon-relay-ng-issue.conf.txt](https://github.com/graphite-ng/carbon-relay-ng/files/974511/carbon-relay-ng-issue.conf.txt) and the two others configuration files with the workaround after 16:30 [carbon-relay-ng-workaround-instance1.conf.txt](https://github.com/graphite-ng/carbon-relay-ng/files/974563/carbon-relay-ng-workaround-instance1.conf.txt) [carbon-relay-ng-workaround-instance2.conf.txt](https://github.com/graphite-ng/carbon-relay-ng/files/974510/carbon-relay-ng-workaround-instance2.conf.txt)
Thank you much for your help.
